text/0000-assoc-default-groups.md

```diff
@@ -0,0 +1,1630 @@
+which is therefore the type of `<Bar as Foo>::QUUX`.
+However, we have not given a different value for the constant,
+and so it must be `42u8`, which has the type `u8`.
+Therefore, we have reached an inconsistency in the type system:
+`<Bar as Foo>::QUUX` is of value `42u8`, but of type `Vec<u8>`.
+So we may accept either `impl Foo for Bar` as defined in (4),
+or the definition of `Foo` as in (2), but not *both*.
+
+[RFC 192] solved this dilemma by rejecting the implementation
+and insisting that if you override *one* associated type,
+then you must override *all* other defaulted items.
+Or stated in its own words:
+
+> + If a trait implementor overrides any default associated types,
+>   they must also override all default functions and methods.
+> + Otherwise, a trait implementor can selectively override individual
+>   default methods/functions, as they can today.
+
+Meanwhile, as we saw in the error message above (3),
+the current implementation takes the alternative approach of accepting
+`impl Foo for Bar` (4) but not the definition of `Foo` as in (2).
+
+## Changes in this RFC
+[changes]: #changes-in-this-rfc
+
+In this RFC, we change the approach in [RFC 192] to the currently implemented
+approach. Thus, you will continue to receive the error message above
+and you will be able to provide associated type defaults.
+
+[specialization]: https://github.com/rust-lang/rfcs/pull/1210
+
+With respect to [specialization], the behaviour is the same.
+That is, if you write (5):
+
+```rust
+#![feature(specialization)]
+
+trait Foo {
+    type Bar;
+
+    fn quux(x: Self::Bar) -> u8;
+}
+
+struct Wibble<T>;
+
+impl<T> Foo for Wibble<T> {
+    default type Bar = u8;
+
+    default fn quux(x: Self::Bar) -> u8 { x }
+}
+```
+
+The compiler will reject this because you are not allowed to assume,
+just like before, that `x: u8`. The reason why is much the same as
+we have previously discussed in the [background].
+
+[current_impl_diverge]: https://play.rust-lang.org/?gist=30e01d77f7045359e30c7d3f3144e984&version=nightly&mode=debug&edition=2015
+
+One place where this proposal diverges from what is currently implemented
+is with respect to the [following example][current_impl_diverge] (6):
+
+```rust
+#![feature(associated_type_defaults)]
+
+trait Foo {
+    type Bar = usize;
+
+    fn baz(x: Self::Bar) -> usize;
+}
+
+impl<T> Foo for Vec<T> {
+    fn baz(x: Self::Bar) -> usize { x }
+}
+```
+
+In the current implementation, (6) is rejected because the compiler will not
+let you assume that `x` is of type `usize`. But in this proposal, you would be
+allowed to assume this. To permit this is not a problem because `Foo for Vec<T>`
+is not further specializable since `baz` in the implementation has not been
+marked as `default`.
+
+### Trait objects
+
+Another divergence in this RFC as compared to the current implementation is
+with respect to trait objects. Currently, if you write (7):
+
+```rust
+trait Foo {
+    type Bar = u8;
```

Comment:
One problem we've encountered in the past around defaults that is perhaps worth pondering: when do we prove that the default is "suitable"?

Example:

```rust
trait Foo<T> {
    type Bar: Clone = Vec<T>;
}
```

Is this an error? We don't *know* that `Vec<T>: Clone`, because we don't know that `T: Clone`. So we could require that the trait declare that `T: Clone` -- on the other hand, it may be that there are useful impls of the trait where `T: Clone` does not hold, but those impls do not use this default.

Reply:
Perhaps this is addressed in the RFC, if so, forgive me, I am skimming quickly.

Reply:
Interesting question / problem... I don't think it's currently addressed in the RFC.

Let's give the name `A` for the set of well-formed programs for which we require that all defaults be well-formed and adhere to the bounds of the trait at the definition site of `Foo<T>`, assuming only the bounds of the trait (namely `Self::Bar: Clone`). IOW, this is the set in which `Foo<T>` as defined above is an error.

Let's give the name `B` for the set of well-formed programs for which we don't require that and check adherence to the bounds of the trait in the `impl`s instead. In this set `Foo<T>` as defined above is WF.

Off-hand, it seems to me that `A ⊆ B` because we get `B` from `A` by loosening a restriction. Thus, it should be possible to start with `A` but extend to `B` if we need to. However, I don't have a good justification for this other than wanting to be conservative. Given that we don't allow other items in `Foo<T>` to assume `Vec<T>` as the default it should be sound.

-------------

Possibly related is the following example ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b10ffc5cf27a33a2dad3af7f44ee1e43)):

```rust
#![feature(associated_type_defaults)]

trait A {
    type B = Self::C;
    type C = Self::B;
}

impl A for () {}

fn _foo() { let _x: <() as A>::B; }

// Removing this function will make the example compile.
fn main() { let _x: <() as A>::B; }
```

Compiling this today will give you (aside: there seems to be a bug here...):

```
error[E0275]: overflow evaluating the requirement `<() as A>::B`

thread '<unnamed>' panicked at 'Metadata module not compiled?', src/libcore/option.rs:1038:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to previous error
```

However, using the set `A`, the program:
```rust
#![feature(associated_type_defaults)]

trait A {
    type B = Self::C;
    type C = Self::B;
}
```
should fail to compile since there is a cycle but it should presumably compile under `B` (as today).